### PR TITLE
add mongo distributed lock

### PR DESCRIPTION
**Pull request by LXPWing**
**What this PR does**:
add mongo distributed lock

**Which issue(s) this PR fixes**:
#104

### CODE REVIEW

**Review by seeflood — CHANGES_REQUESTED** [review]
`demo/configuration/apollo/apolloClientDemo`  
There is no need to submit this binary file  :)

**Comment by seeflood** [comment]
关于如何在ut中mock掉mongoDB:
我看mongo的sdk是返回一个叫 `mongo.Client`的struct对吧。struct不好mock，所以一般写ut的关键是：
- 先自己把struct封装成一个interface。具体来说，先定义一个interface，然后你的`NewMongoClient(m MongoMetadata)`这个函数应该返回这个interface，而不是`mongo.Client`这个struct
- 一但封装成interface了，在ut里就可以把他替换成自己的mock实现了

可以参考 @ZLBer 之前写的[zookeeper lock](https://github.com/mosn/layotto/blob/main/components/lock/zookeeper/zookeeper_lock.go)

他当时也面临同样的问题（如何在ut里mock zk),解决方案是把zk 的connection封装成一个interface:
https://github.com/mosn/layotto/blob/e0e5eba832ba5619134d863a51a4591a66ed734e/components/pkg/utils/zookeeper.go#L53

**Comment by seeflood** [comment]
@ZLBer @whalesongAndLittleFish @stulzq Hi guys, could u help review this PR?

File: components/lock/mongo/mongo_lock.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+	status, err := session.WithTransaction(e.ctx, func(sessionContext mongo.SessionContext) (interface{}, error) {
+		var status int
+
+		result, err := e.collection.DeleteOne(e.ctx, bson.M{"_id": req.ResourceId, "LockOwner": req.LockOwner})
+
+		if result.DeletedCount == 1 && err == nil {
+			status = 0
+		} else if result.DeletedCount == 0 && err == nil {
+			if cursor, err := e.collection.Find(e.ctx, bson.M{"_id": req.ResourceId}); cursor != nil && cursor.RemainingBatchLength() != 0 && err == nil {
+				status = 2
```
Comment (ZLBer) [inline components/lock/mongo/mongo_lock.go:171]: There is a situation:  another lock  time expired,  but it did not  release the lock(may be down), parameter expireTime  is not used？

Comment (LXPWing) [reply components/lock/mongo/mongo_lock.go:171]: @ZLBer Thanks for the review,I will use both Chinese and English for easy communication
Deadlock, there are 3 possible cases on the server side
1. The lock has not expired 
2. The lock has expired and has been released. 
3. lock has expired due to mongodb's ttl mechanism (default is 60s to perform a clear) resulting in lock not cleared
If it is the third case I think the user should set the ttl execution time

@ZLBer 谢谢你的review
为了方便交流我采用中英两种语言
死锁，服务端可能存在3中情况
1.锁未过期 
2.锁已过期并被清除
3.锁已过期由于mongodb的ttl机制(默认为60s执行一次清除)导致锁未清除
如果是第三种情况我认为应该用户去设置ttl清除执行时间

Comment (LXPWing) [reply components/lock/mongo/mongo_lock.go:171]: @ZLBer mongo的ttl有两种写法expireAfterSeconds是否为0
1.expireAfterSeconds不为0时，例子：createIndex( { "createdAt": 1 }, { expireAfterSeconds: 3600 } )创建过期索引设置createdAt为过期字段并在插入文档后的36000s过期清除
2.expireAfterSeconds为0时，例子：createIndex( { "createdAt": 1 }, { expireAfterSeconds: 0 } )创建过期索引设置createdAt为过期字段（过期方式时按createdAt的时间过期，例如“createdAt”: "2021.12.7.2:20" 在2021.12.7.2:20的时候这个文档过期清除）
mongo的清除动作是mongo数据库后台启动一个异步线程清除的（这个线程启动是60s一次，用户可以根据需求去设置这个异步线程的执行频率），所以就会产生过期数据没被删除（这里无法确保数据100%正确）。我在代码层面只是使用了事务和强一致读写，并没有设置这个ttl清除的执行频率（这里频率不好控制具体还是要看场景）
我这里的思路是先删除锁，然后根据删除的结果去判断锁的状态

Comment (ZLBer) [reply components/lock/mongo/mongo_lock.go:171]: @LXPWing  ohhhh  我明白了  一开始的时候创建了个Expire为index的collection, 用这个字段来保存过期时间，并设置SetExpireAfterSeconds(0)。可以在代码里多加点注释~

File: components/lock/mongo/mongo_lock.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+// Features is to get MongoLock's features
+func (e *MongoLock) Features() []lock.Feature {
+	return e.features
+}
+
+func (e *MongoLock) TryLock(req *lock.TryLockRequest) (*lock.TryLockResponse, error) {
+	var err error
+	// create mongo session
+	e.session, err = e.client.StartSession()
```
Comment (ZLBer) [inline components/lock/mongo/mongo_lock.go:126]: check error

File: components/lock/mongo/mongo_lock.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+		return &lock.TryLockResponse{
+			Success: false,
+		}, nil
+	}
+}
+
+func (e *MongoLock) Unlock(req *lock.UnlockRequest) (*lock.UnlockResponse, error) {
+	var err error
+	// create mongo session
+	e.session, err = e.client.StartSession()
```
Comment (ZLBer) [inline components/lock/mongo/mongo_lock.go:184]: check error

File: components/lock/mongo/mongo_lock.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+		// set exprie date
+		expireTime := time.Now().Add(time.Duration(req.Expire) * time.Second)
+
+		// find mongo lock
+		singleResult = e.collection.FindOne(e.ctx, bson.M{"_id": req.ResourceId})
+
+		// insert mongo lock
+		if singleResult.Err() == mongo.ErrNoDocuments {
+			insertOneResult, err = e.collection.InsertOne(e.ctx, bson.M{"_id": req.ResourceId, "LockOwner": req.LockOwner, "Expire": expireTime})
+		}
```
Comment (ZLBer) [inline components/lock/mongo/mongo_lock.go:148]: Is it necessary to find first and then insert ？ only insert  and if  `_id` exsit ,  mongo will return error？

Comment (LXPWing) [reply components/lock/mongo/mongo_lock.go:148]: yes, primary key unique

Comment (ZLBer) [reply components/lock/mongo/mongo_lock.go]: so the transaction maybe not necessary ？

**Comment by ZLBer** [comment]
@LXPWing  no big problem, next ,you need to add docs in `docs/_sidebar.md` `docs/en/component_specs/lock/XX.md` `docs/zh/component_specs/lock/XX.md`.by the way ,the mock can use gomock, it‘s quick. 
thanks for your contribution.

**Review by whalesongAndLittleFish — APPROVED** [review]
nice~,LGTM

**Comment by seeflood** [comment]
Sorry for not replying !
I found that there are many ut failures in component package (including some compilation errors)
I will submit a PR to fix the test cases and then go back to test and merge this PR.
Thanks!

**Review by zhenjunMa — APPROVED** [review]

**Review by seeflood — APPROVED** [review]

**Comment by seeflood** [comment]
Thanks again for your contribution!
This component will be included in layotto v0.4

**Comment by ZLBer** [comment]
@LXPWing Thanks for your contribution! Maybe you can try another community task~

**Pull request merged**